### PR TITLE
First pass at public API

### PR DESCRIPTION
**Pull request by alanmcruickshank**
Relates to #289 (although doesn't yet provide the full functionality suggested by @barrywhart yet).
It provides a basic outline as required for: https://github.com/sqlfluff/sqlfluff-online/issues/14

This also reorganises the repo & docs in advance of the additional functionality planned for the rest of the public API.
- Most of the core sqlfluff library has been moved into `sqlfluff.core` with `sqlfluff.cli`, `sqlfluff.api` and `sqlfluff.diff_cover_plugin` referencing it.
- An `examples` folder in the repo to cover example code using `sqlfluff` as an `import`.
- An `API Reference` section of the docs with examples and outlines for the public methods so far.
- A big update a all the references and imports to support the new structure.
- A little bit of syntactic sugar in instantiating `sqlfluff.Linter()` to dedupe some sections of code when no complex config is required.
- There's also a small css change to the docs to make them more readable in the autodoc sections.

Would appreciate feedback from @nolanbconaway in particular as to how this might work in combination with sqlfluff-online. Also from @barrywhart on how the internal structuring acts as a stepping stone to what we want to achieve in #289.

Feedback from others on the restructure also very welcome!

Next step would be to either: clean up the `Linter()` class for public usage or expose a `sqlfluff.parse()` method which returns something usable to infer structure.

### CODE REVIEW

File: examples/basic_api_usage.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+"""This is an exmaple of how to use the simple sqlfluff api."""
```
Comment (barrywhart) [inline examples/basic_api_usage.py:1]: Typo: exmaple

File: examples/basic_api_usage.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+#     {"code": "L010", "line_no": 1, "line_pos": 1, "description": "Inconsistent capitalisation of keywords."}
+#     ...
+# ]
+
+#  -------- FIXING ----------
+
+# Fix the given string and get a string back which has been fixed.
+result = sqlfluff.fix(my_bad_query, dialect="bigquery")
+
+# result = 'SELECT  *, 1, blah AS  foo  FROM mytable\n'
```
Comment (barrywhart) [inline examples/basic_api_usage.py:23]: There are multiple spaces after `SELECT` and `AS`, also multiple spaces before `FROM`. Is this the actual/intended behavior?

File: src/sqlfluff/api/simple.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+"""The simple public API methods."""
+
+from ..core import Linter
+
+
+def _unfiy_str_or_file(sql):
```
Comment (barrywhart) [inline src/sqlfluff/api/simple.py:6]: Typo in the function name: `_unfiy` -> `_unify`

File: src/sqlfluff/api/simple.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    sql = _unfiy_str_or_file(sql)
+    linter = Linter(dialect=dialect)
+
+    result = linter.lint_string_wrapped(sql)
+    result_records = result.as_records()
+    # Return just the violations for this file
+    return result_records[0]["violations"]
+
+
+def fix(sql, dialect="ansi"):
```
Comment (barrywhart) [inline src/sqlfluff/api/simple.py:37]: Should we consider letting the caller specify which fixes to apply?

File: src/sqlfluff/api/simple.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    """Unify string and files in the same format."""
+    if not isinstance(sql, str):
+        try:
+            sql = sql.read()
+        except AttributeError:
+            raise TypeError("Value passed as sql is not a string or a readable object.")
+    return sql
+
+
+def lint(sql, dialect="ansi"):
```
Comment (barrywhart) [inline src/sqlfluff/api/simple.py:16]: Should we consider letting the caller specify which rules to check?

File: src/sqlfluff/core/linter.py [hunk 6]
```diff
@@ -564,9 +573,19 @@ def persist_changes(self, formatter=None, **kwargs):
...
     """The interface class to interact with the linter."""
 
-    def __init__(self, sql_exts=(".sql",), config=None, formatter=None):
-        if config is None:
-            raise ValueError("No config object provided to linter!")
+    def __init__(self, sql_exts=(".sql",), config=None, formatter=None, dialect=None):
+        if dialect and config:
+            raise ValueError(
+                "Cannot specify `dialect` and `config`. Any config object "
+                "specifies it's own dialect."
```
Comment (barrywhart) [inline src/sqlfluff/core/linter.py:580]: Nitpick: No apostrophe is needed in possessive "its": "it's" is short for "it is".

File: test/api/simple_test.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
+"""Tests for simple use cases of the public api.
+
+These tests should also test the imports, not just the functionality.
+"""
+
+
+def test__api__lint_string():
+    """Basic checking of lint functionality."""
+    import sqlfluff
```
Comment (barrywhart) [inline test/api/simple_test.py:9]: Usually, imports are at file (module) scope. Is there a reason these tests are `import`-ing inside the function?

**Review by barrywhart — CHANGES_REQUESTED** [review]

File: docs/source/_static/custom.css [hunk 8]
```diff
@@ -0,0 +1,6 @@
+/* Overrides for spacing within autodoc */
+
+dl.py {
+    /* Spacing is normally far too tight, this makes a bit more spact between elements */
```
Comment (pwildenhain) [inline docs/source/_static/custom.css:4]:
```suggestion
    /* Spacing is normally far too tight, this makes a bit more space between elements */

```

**Review by pwildenhain — COMMENTED** [review]
Great start. 💯 agree with @barrywhart about allowing user to configure lints/fixes

Seems like this also allows all fixes to be applied at once -- so this also _sort of_ related to #334

File: examples/basic_api_usage.py (hunk 2 above)
Comment (alanmcruickshank) [reply examples/basic_api_usage.py:23]: That's an excellent point! There is currently no linting rule to enforce single whitespace outside of indents and comments, but I think it's a great idea. I'll make an issue for it.

File: examples/basic_api_usage.py (hunk 1 above)
Comment (alanmcruickshank) [reply examples/basic_api_usage.py:1]: 🤦

File: src/sqlfluff/api/simple.py (hunk 3 above)
Comment (alanmcruickshank) [reply src/sqlfluff/api/simple.py:6]: 🤦

File: src/sqlfluff/api/simple.py (hunk 5 above)
Comment (alanmcruickshank) [reply src/sqlfluff/api/simple.py:16]: yes. I'll add that. I assume the choice would be by code reference, so I'll do that first for the next round of review.

File: src/sqlfluff/core/linter.py (hunk 6 above)
Comment (alanmcruickshank) [reply src/sqlfluff/core/linter.py:580]: 🤦

File: test/api/simple_test.py (hunk 7 above)
Comment (alanmcruickshank) [reply test/api/simple_test.py:9]: I was on the fence about this one, with the initial intent being to also *test the import*, but I guess we test that anyway as part of the test suite. I think I was trying to be too clever - and will revert to standard.

File: src/sqlfluff/api/simple.py (hunk 4 above)
Comment (alanmcruickshank) [reply src/sqlfluff/api/simple.py:37]: similar to the rules one, I'll add that in. 👍

File: examples/basic_api_usage.py (hunk 2 above)
Comment (alanmcruickshank) [reply examples/basic_api_usage.py:22]: See #502

**Comment by alanmcruickshank** [comment]
## Update to this PR. 

I've also added in a `sqlfluff.parse` command which is a bit more advanced, but starts to get to the really useful use cases brought up in #289 . The last piece of this puzzle is to be able to parse out table references or column references, which is probably going to need some more detailed work. At the moment both column references, table references and references to subqueries or CTEs all get parsed as `object_reference`, so we'll need to introduce some changes to the parser to facilitate that [and that's a task for another day].

File: examples/basic_api_usage.py (hunk 2 above)
Comment (nolanbconaway) [reply examples/basic_api_usage.py:22]: wow barry good eyes!

**Review by nolanbconaway — APPROVED** [review]
Alan! This is _choice_ 👌 . It'll be very straightforward to shift the subprocess hack in sqlfluff-online to this api.

I have two comments, neither of which should block this merge. These are flags for future functionality that come naturally from this interface:

1. Not all relevant CLI args are available for use in the API (mainly: `templater`, `exclude-rules`).
2. All this config may come from `.sqlfluff` or other config file. Maybe this is already done under the hood on the call to `Linter()`  (ignore this if so) but if not, the public API ought to eventually read the same config as the CLI.

**Comment by alanmcruickshank** [comment]
Thanks @nolanbconaway !

As for your future comments, my intent is that the api as described will only support a subset of the functionality that's available. For more complex use cases, people should interact with `sqlfluff.Linter()` directly, and then will have access to all the config options you mention. To support that properly, I think that the `Linter` class needs some TLC first, but at least this is a step in the right direction. Very happy to have issues raised to prompt which bits we do first as and when people come up against those use cases.

**Pull request merged**